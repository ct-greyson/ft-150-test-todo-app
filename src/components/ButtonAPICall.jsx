import React from "react";
import axios from "axios";

const ButtonAPICall = () => {
  const fetchTodos = async () => {
    try {
      const response = await axios.get(
        "https://jsonplaceholder.typicode.com/todos"
      );
      console.log(response.data);
    } catch (error) {
        console.log(error.message)
    }
  };

  return (
    <div>
      <h1>woah, it's a button</h1>
      <button onClick={fetchTodos}>Fetch Todos</button>
    </div>
  );
};

export default ButtonAPICall;
