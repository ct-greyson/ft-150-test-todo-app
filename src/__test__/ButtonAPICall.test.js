import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import axios from "axios";
import ButtonAPICall from "../components/ButtonAPICall";

// create a fake version (a "mock" version) of our axios so we don't actually make API calls
jest.mock("axios");

// describe allows us to name a group of our tests and bundle them together into a block
// takes in a string describing what the tests are for, and a function to perform the tests
describe("ButtonAPICall Component Tests", () => {
  // block for the test itself
  // name should be specific/concise
  test("fetches data from API when button is clicked", async () => {
    // mock data to return from API
    const mockResponse = { data: [{ id: 1, title: "Buy food" }] };

    // mock the API call to return our mockResponse
    // when API is called, this is the call that will be made
    axios.get.mockResolvedValue(mockResponse)

    // first, we render our component to the fake DOM
    // we search for an element in that component with the text "Fetch Todos" (case sensitive) 
    // then we trigger a click event on that element
    const { getByText } = render(<ButtonAPICall />)
    fireEvent.click(getByText("Fetch Todos"))

    // what is the event supposed to do
    // waitFor waits for the event to be fired
    await waitFor(() => {
        // this is the API we are expecting to call with axios when the event has been triggered
        expect(axios.get).toHaveBeenCalledWith("https://jsonplaceholder.typicode.com/todos")
    })
  });

  //Snapshot/UI test
  test("Component matches snapshot", () => {
    // asFragment creates the snapshot for our code if one does not exist
    // it captures the current state of our UI that will be compared against for future testing
    // if a snapshot already exists, then we go straight to the comparison
    const { asFragment } = render(<ButtonAPICall />)
    expect(asFragment()).toMatchSnapshot(); //compares the current snapshot to the original 
  })
});


