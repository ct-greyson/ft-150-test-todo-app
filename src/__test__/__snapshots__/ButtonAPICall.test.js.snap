// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonAPICall Component Tests Component matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      woah, it's a button
    </h1>
    <button>
      Fetch Todos
    </button>
  </div>
</DocumentFragment>
`;
